streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -744,21 +740,18 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
             innerFailure    <- Promise.make[Cause[E1], Nothing].toManaged_
             interruptInners <- Promise.make[Nothing, Unit].toManaged_
             cancelers       <- Queue.bounded[Promise[Nothing, Unit]](n).toManaged(_.shutdown)
-            cancelersGuard  <- Semaphore.make(1).toManaged_
             _ <- self.foreachManaged { a =>
                   for {
-                    latch    <- Promise.make[Nothing, Unit]
                     canceler <- Promise.make[Nothing, Unit]
+                    latch    <- Promise.make[Nothing, Unit]
+                    size     <- cancelers.size
```

Comment:
I liked `cancelersGuard` but simpler is of course better :-) Are there any race conditions that could happen here with the handlers in the `foldCauseM` or the finalizer on the outer stream below?

Reply:
Technically, and I realized this when developing this PR, the `cancelers` queue is protected by the `permits` semaphore, if we block the continutaion of the `foreachManaged`, which we have `Managed` to do here. Sorry for the bad joke.

Reply:
Again, the `latch` makes sure that an inner stream gets at least a chance to run its `bracket` operation, which means that the finalizer on the outer stream has to compete with the last `n` inner streams for the `permits`, at which point this becomes semantically the same with `flatMapPar`, so I'd think that we are pretty safe from any race conditions.